### PR TITLE
add support NNUE, Threads; change piece buttons to square

### PR DESCRIPTION
**Pull request by NelloHo**

### CODE REVIEW

File: fairyfishgui.py [hunk 1]
```diff
@@ -177,7 +178,7 @@ def square2idx(self, square):
     def render_square(key, location):
         square_color = SQUARE_COLORS[(location[0] + location[1]) % 2]
         font_size = min(sg.Window.get_screen_size()) // 50
-        button = sg.Button(size=(3, 2), button_color=square_color, pad=(0, 0), font='Any {}'.format(font_size), key=key)
+        button = sg.Button(size=(4, 2), button_color=square_color, pad=(0, 0), font='Any {}'.format(font_size), key=key)
```
Comment (ianfab) [inline fairyfishgui.py]: For me the layout looks better with the previous size. I think this depends on screen resolution, so it is a bit tricky to make it look good independent of screen resolution with this framework. I would suggest to split this topic out into a separate PR/issue as it seems rather independent.

File: fairyfishgui.py [hunk 2]
```diff
@@ -213,14 +214,15 @@ def __init__(self):
...
-                        [sg.Button('Load engine', key='_engine_'), sg.Button('Engine on/off', key='_toggle_')],
-                        [sg.Button('Set FEN', key='_set_fen_'), sg.Button('Reset', key='_reset_')],
-                        [sg.Button('Move', key='_move_'), sg.Button('Undo', key='_undo_')],
-                        [sg.Text('Move List')],
-                        [sg.Multiline(do_not_clear=True, autoscroll=True, size=(25, 10), key='_movelist_')],
-                        [sg.Text('Engine Output')],
-                        [sg.Multiline(do_not_clear=True, autoscroll=True, size=(25, 10), key='_engine_output_')],
-                        ]
+                          [sg.Button('Load engine', key='_engine_'), sg.Button('Engine on/off', key='_toggle_')],
+                          [sg.Button('Load NNUE', key='_nnue_'), sg.Button('Set Threads',key='_thread_')],
```
Comment (ianfab) [inline fairyfishgui.py]: Long-term I would suggest to make an engine settings dialogue accessible via the menu bar and/or a button. However, for now I would also be fine with this solution, so I leave it up to you if you want to try to change it.

File: fairyfishgui.py [hunk 3]
```diff
@@ -321,13 +323,26 @@ def update_board(self, variant=None, fen=None, move=None, undo=False):
...
             if self.engine:
-                self.engine.position(self.board.state.start_fen, self.board.state.moves)
+                self.engine.position(
+                    self.board.state.start_fen, self.board.state.moves)
         if undo:
             self.board.state.pop()
             if self.engine:
                 self.engine.position(self.board.state.start_fen, self.board.state.moves)
+				
+        if nnue:
```
Comment (ianfab) [inline fairyfishgui.py]: Setting the options should be independent from updating the board state, so I would suggest to either use `self.engine.setoption` directly from within the main loop or wrap it into another method (`self.set_engine_option` or the like). Stopping and restarting the engine search as part of that makes sense, but resetting the position seems unnecessary to me.

File: fairyfishgui.py [hunk 4]
```diff
@@ -373,7 +387,25 @@ def run(self):
...
                 if move:
                     self.update_board(move=move)
-
+            elif button == '_nnue_':
+                if not self.engine:
+                    sg.popup_ok('Please load the engine first')
+                else:
+                    nnue_path = sg.popup_get_file('Select NNUE',file_types=(('nnue file', '*.nnue'),))
+                    if nnue_path:
+                        if not (nnue_path.split('/')[-1].startswith(self.board.state.variant) if self.board.state.variant != 'chess' else nnue_path.split('/')[-1].startswith('nn')):
```
Comment (ianfab) [inline fairyfishgui.py]: For safely splitting path components independent of the OS (`/` vs `\`) there is `os.path.split`, or in this case more specifically `os.path.basename`. Very minor change, but looks much cleaner.

This check anyway should not really be necessary though as the engine won't complain about "wrong" NNUE files, it will just only use them once the corresponding variant is loaded, so it is safe either way. If it is meant as a notification for the user to prevent him from accidentally providing the wrong NNUE file it would need to be more sophisticated as it currently also would prevent valid usage of NNUEs, e.g., using a makruk net for ouk (see nnueAlias).

**Review by ianfab — CHANGES_REQUESTED** [review]
Thanks for your PR. I added a few CR comments, see below.

**Comment by NelloHo** [comment]
Thanks for the reviewing !
Sure, I will fix all what you said.

File: fairyfishgui.py [hunk 5]
```diff
@@ -33,7 +32,7 @@ def setoption(self, name, value):
     def initialize(self):
         message = 'uci\n'
         for option, value in self.options.items():
-            message += self.process.stdin.write('setoption name {} value {}\n'.format(option, value))
```
Comment (NelloHo) [inline fairyfishgui.py:36]: I change this line cz I think it might be what you mean.
Have changed all the reviews as you said

File: fairyfishgui.py [hunk 6]
```diff
@@ -298,11 +320,29 @@ def format_info(info):
...
 
+    def set_engine_option(self, nnue=None, threads=None):
+        if self.engine and not self.engine.paused and (nnue or threads):
+            self.engine.stop()
+            self.engine.paused = False
+        if nnue:
+            if self.engine:
+                self.engine.setoption('EvalFile',nnue)
+            else:
+                self.engine_settings['EvalFiles'] = nnue
```
Comment (ianfab) [inline fairyfishgui.py]: Thanks for your updates. I think there is just one typo here, otherwise looks good to me.

**Review by ianfab — APPROVED** [review]
